src/commands/run.rs

```diff
@@ -13,24 +25,122 @@ impl Run {
         Self {}
     }
 
+    fn is_executable_in_path(program: &str) -> bool {
+        if let Ok(path) = std::env::var("PATH") {
+            for p in path.split(':') {
+                let p_str = format!("{}/{}", p, program);
+                if fs::metadata(p_str).is_ok() {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
     pub fn exec(
         &self,
         input_dir: String,
         port: u32,
         filter_search_domain: String,
+        serve: bool,
+        rootless: bool,
     ) -> Result<(), Error> {
-        debug!(
-            "Setting up aardvark server with input directory as {:?}",
-            input_dir
-        );
-
-        if let Err(er) = serve::serve(&input_dir, port, &filter_search_domain) {
-            return Err(std::io::Error::new(
-                std::io::ErrorKind::Other,
-                format!("Error starting server {}", er),
-            ));
+        let current_binary_path = match std::env::current_exe() {
+            Ok(path) => path.to_str().unwrap().to_string(),
+            Err(e) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("failed to get current binary path: {}", e),
+                ));
+            }
+        };
+
+        if serve {
+            debug!(
+                "Setting up aardvark server with input directory as {:?}",
+                input_dir
+            );
+
+            println!("current binary path{:?}", std::env::current_exe());
+
+            if let Err(er) = serve::serve(&input_dir, port, &filter_search_domain) {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("Error starting server {}", er),
+                ));
+            }
+            Ok(())
+        } else {
+            // double fork and verify if server is running
+            // and exit parent
+            // Why double fork ?
+            // Its important that nature of aardvark server is more like a daemon
+            // so following block ensures that process returns when aardvark server
+            // is ready to serve
+            //
+            // setsid() ensures that there is no controlling terminal on the child process
+
+            match unsafe { fork() } {
+                Ok(ForkResult::Parent { child, .. }) => {
+                    log::debug!("starting aardvark on a child with pid {}", child);
+                    if let Err(err) = waitpid(Some(child), None) {
+                        log::debug!("error while waiting for child pid {}", err);
+                    }
+                    // verify aardvark here
```

Comment:
Must create a dummy DNS request to verify if server is running before quitting parent.

Reply:
The problem is call to server is blocking , either we start server in a separate thread or fork and start the actual server.

Reply:
Since we already dicussed this on IRC following block stays and parent verifies if child is ready before quitting and returning success to netavark.